### PR TITLE
[USH-3692] Add datadog metrics for dynamic search

### PR DESCRIPTION
**Pull request by akashkj**
## Product Description
Add metrics to capture processing time for:
1. API call to do query search in formplayer
2. Server to perform case search

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3692

[DD dashboard](https://app.datadoghq.com/dashboard/6f7-5dt-fvb?refresh_mode=sliding&from_ts=1699297755723&to_ts=1699301355723&live=true)

image

## Feature Flag
DYNAMICALLY_UPDATE_SEARCH_RESULTS

### Safety story
Tested on staging. No change in the existing functionality, added additional logs to record metrics.
Also added dashboard in datadog with relevant metrics to check that API calls are captured correctly.

### Automated test coverage
Not added

### QA Plan
Not planned

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/util/metrics/datadog.py [hunk 1]
```diff
@@ -61,6 +61,13 @@ def _gauge(self, name: str, value: float, tags: Dict[str, str] = None, documenta
         dd_tags = _format_tags(tags)
         _datadog_record(statsd.gauge, name, value, dd_tags)
 
+    def _datadog_histogram(self, metric_name, metric_value):
+        """
+        https://docs.datadoghq.com/metrics/types/?tab=histogram#metric-types
+        """
+        print(f"Logging {metric_name} - {metric_value}")
+        _datadog_record(statsd.histogram, metric_name, metric_value)
```
Comment (millerdev) [inline corehq/util/metrics/datadog.py:69]: FYI, histogram stats in datadog are only useful in very specific scenarios because the aggregation window is a constant defined on the client, which means the data will not be useful unless you have a meaningful sample of events consistently recorded within that time period. Unfortunately most of the links I used in the past are now broken, but some discussion on this topic can be found [here](https://github.com/dimagi/commcare-hq/pull/17030#discussion_r127495077) and [here](https://github.com/dimagi/commcare-hq/pull/17080#issuecomment-316395782).

Comment (millerdev) [reply corehq/util/metrics/datadog.py:69]: We have implemented a different histogram recording technique, which uses counter buckets. You can read more about it here:

https://github.com/dimagi/commcare-hq/blob/4d7b39cc6f8438874a66745136aa67ef75ea9da6/corehq/util/metrics/datadog.py#L64-L90

Comment (akashkj) [reply corehq/util/metrics/datadog.py:69]: Thank you, I'll read through these references. Will try to use existing 'histogram' metric instead of adding datadog's native histogram.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js [hunk 2]
```diff
@@ -209,6 +209,9 @@ hqDefine("cloudcare/js/formplayer/router", function () {
             // force manual search in split screen case search for workflow compatibility
             sidebarEnabled ? { forceManualSearch: true } : {}
         );
+
+        var queryObject = _.extend(queryObject, initiator === "dynamicSearch" ? { initiatedBy: "dynamic_search"} : {});
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js:213]: Why the change from camel case to snake case? Couldn't this be:
```suggestion
        var queryObject = _.extend(queryObject, {initiatedBy: initiator});
```

Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js]: Changed

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ hqDefine("cloudcare/js/formplayer/menus/api", function () {
         FormplayerFrontend = hqImport("cloudcare/js/formplayer/app"),
         formplayerUtils = hqImport("cloudcare/js/formplayer/utils/utils"),
         ProgressBar = hqImport("cloudcare/js/formplayer/layout/views/progress_bar");
```
Comment (stephherbers) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js:13]: change ending to a comma to get rid of those errors

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js [hunk 4]
```diff
@@ -209,6 +209,9 @@ hqDefine("cloudcare/js/formplayer/router", function () {
             // force manual search in split screen case search for workflow compatibility
             sidebarEnabled ? { forceManualSearch: true } : {}
         );
+
+        var queryObject = _.extend(queryObject, {initiatedBy: initiator});
```
Comment (stephherbers) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js:213]: don't need to initialize `queryObject` again just to use `_.extend`

Comment (esoergel) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js:213]: also I think you can just stick this in the initial definition after line 207

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js [hunk 5]
```diff
@@ -264,15 +264,23 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
             this.sortIndex = null;
         };
 
-        this.setQueryData = function ({ inputs, execute, forceManualSearch}) {
+        this.setQueryData = function ({ inputs, execute, forceManualSearch, initiatedBy}) {
```
Comment (esoergel) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:267]: I notice this function calls it `initiatedBy`, but others call it `initiator` - could you unify the terminology for clarity?

File: corehq/apps/cloudcare/views.py [hunk 6]
```diff
@@ -605,3 +607,24 @@ def _fail(error):
...
+@login_and_domain_required
+@require_POST
+def api_histogram_metrics(request, domain):
+    request_dict = request.POST
+
+    metric_name = request_dict.get("metrics")
+    try:
+        duration = float(request_dict.get("responseTime"))
+    except ValueError:
+        duration = None
```
Comment (esoergel) [inline corehq/apps/cloudcare/views.py:621]: shouldn't this be treated as an error?

Comment (akashkj) [reply corehq/apps/cloudcare/views.py:621]: This is added as a fail-safe but if checked, from frontend data will always be sent as a float string containing duration in ms as done [here](https://github.com/dimagi/commcare-hq/pull/33678/files/c933dc44eacb59e296879abb49981c1e1e582dcc#diff-e9a325527828d83ca273a6ebef7ca67882d9dee8484486bbfcea88215b8c8531R172). 
Should I remove this try/except instead as ideally this exception will not happen?

Comment (esoergel) [reply corehq/apps/cloudcare/views.py:621]: I think so, then we'd get a sentry notification if this broke down the road.

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js (hunk 3 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/api.js]: Changed

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js (hunk 4 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js]: Moved to initial declaration

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js (hunk 5 above)
Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:267]: Changed to `initiatedBy` at other places

File: corehq/apps/cloudcare/views.py (hunk 6 above)
Comment (akashkj) [reply corehq/apps/cloudcare/views.py]: Removed try/except

File: corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js [hunk 7]
```diff
@@ -564,13 +564,14 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             self.performSubmit();
```
Comment (stephherbers) [inline corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:564]: what about this usage of performSubmit() with the param change?

Comment (akashkj) [reply corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:564]: This believe this should be fine as `initiatedBy` will be `undefined` in this call, which will be handled [here](https://github.com/dimagi/commcare-hq/pull/33678/files/84a85c89f8e8f82951a7f79b5012e75d9bf770fc#diff-8b8c7bc0774737538270c8660735bf9f33873d5295f70efc4775d86487f01d74R278)

**Review by stephherbers — APPROVED** [review]

**Comment by akashkj** [comment]
@stephherbers Could you please approve it again. New change is a linting fix.

**Review by stephherbers — APPROVED** [review]
